File: gtci/02-fast-and-slow-pointers/cpp/5-palindrome-linked-list.cpp
#include <cassert>
#include <iostream>

using namespace std;

class ListNode {
   public:
    int val = 0;
    ListNode* next;

    ListNode(int value) {
        this->val = value;
        next = nullptr;
    }

    ListNode(int value, ListNode* next) {
        this->val = value;
        this->next = next;
    }
};

class Solution {
   public:
    bool isPalindrome(ListNode* head) {
        ListNode* middle = Solution::getMiddleNode(head);

        ListNode* firstHalfStart = Solution::reverseList(head);
        ListNode* secondHalfStart = Solution::reverseList(middle);

        while (firstHalfStart != nullptr and secondHalfStart != nullptr) {
            if (firstHalfStart->val != secondHalfStart->val) return false;

            firstHalfStart = firstHalfStart->next;
            secondHalfStart = secondHalfStart->next;
        }

        Solution::reverseList(secondHalfStart);

        return true;
    }

   private:
    static ListNode* getMiddleNode(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        return slow;
    }

    static ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;
        while (curr != nullptr) {
            ListNode* temp = curr->next;

            curr->next = prev;

            prev = curr;
            curr = temp;
        }

        return prev;
    }
};

int main() {
    Solution sol;

    ListNode* head = new ListNode(
        1, new ListNode(2, new ListNode(3, new ListNode(2, new ListNode(1)))));
    ListNode* head2 = new ListNode(
        1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    ListNode* head3 =
        new ListNode(1, new ListNode(2, new ListNode(2, new ListNode(1))));

    assert(sol.isPalindrome(head) == true);
    assert(sol.isPalindrome(head2) == false);
    assert(sol.isPalindrome(head3) == true);

    cout << "All test cases passed." << endl;

    // Clean up memory
    for (ListNode* curr = head; curr != nullptr;) {
        ListNode* temp = curr;
        curr = curr->next;
        delete temp;
    }

    return 0;
}

File: gtci/02-fast-and-slow-pointers/python3/5-palindrome-linked-list.py
class Node:
    def __init__(self, value, next=None):
        self.val = value
        self.next = next


class Solution:
    def isPalindrome(self, head):
        middle = self.get_middle_node(head)

        end = self.reverse_list(middle)

        l1 = head
        l2 = end
        while l1 is not None and l2 is not None:
            if l1.val != l2.val:
                return False

            l1 = l1.next
            l2 = l2.next

        self.reverse_list(end)

        return True

    def get_middle_node(self, head):
        slow = head
        fast = head
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next

        return slow

    def reverse_list(self, head):
        prev = None
        curr = head
        while curr is not None:
            temp = curr.next

            curr.next = prev
            prev = curr
            curr = temp

        return prev


if __name__ == "__main__":
    sol = Solution()

    head = Node(1, Node(2, Node(3, Node(2, Node(1)))))
    head2 = Node(1, Node(2, Node(3, Node(4, Node(5)))))
    head3 = Node(1, Node(2, Node(2, Node(1))))

    assert sol.isPalindrome(head) == True
    assert sol.isPalindrome(head2) == False
    assert sol.isPalindrome(head3) == True

    print("All test cases passed.")
